gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchExecutor.java

```diff
@@ -0,0 +1,147 @@
+import com.google.cloud.hadoop.util.GoogleCloudStorageEventBus;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.io.IOException;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * BatchExecutor provides a means to manually batch requests using a thread pool. Execution is
+ * performed by the underlying {@link #requestsExecutor} ExecutorService.
+ *
+ * <p>Expected usage is to create a new BatchExecutor instance per client operation that represents
+ * logical grouping of requests(delete, copy, get).
+ *
+ * <p>Instance of this class can not be used again after {@link #shutdown()} method has been called.
+ */
+class BatchExecutor {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private final ExecutorService requestsExecutor;
+  private final Queue<Future<Void>> responseFutures = new ConcurrentLinkedQueue<>();
+
+  public BatchExecutor(int numThreads) {
+    this.requestsExecutor =
+        numThreads == 0 ? newDirectExecutorService() : newRequestExecutor(numThreads);
+  }
+
+  private static ExecutorService newRequestExecutor(int numThreads) {
+    ThreadPoolExecutor executor =
+        new ThreadPoolExecutor(
+            /* corePoolSize= */ numThreads,
+            /* maximumPoolSize= */ numThreads,
+            /* keepAliveTime= */ 10L,
+            TimeUnit.SECONDS,
+            new LinkedBlockingQueue<>(numThreads * 20),
+            new ThreadFactoryBuilder()
+                .setNameFormat("gcs-manual-batching-pool-%d")
+                .setDaemon(true)
+                .build());
+    executor.allowCoreThreadTimeOut(true);
+    executor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+    return executor;
+  }
+
+  /** Adds a task to the execution queue. */
+  public <T> void queue(Callable<T> task, JsonBatchCallback<T> callback) {
+    checkState(
+        !requestsExecutor.isShutdown() && !requestsExecutor.isTerminated(),
+        "requestExecutor should not be terminated to queue request");
+
+    responseFutures.add(
+        requestsExecutor.submit(
+            () -> {
+              execute(task, callback);
+              return null;
+            }));
+  }
+
+  private static <T> void execute(Callable<T> task, JsonBatchCallback<T> callback)
+      throws Exception {
+    try {
+      T res = task.call();
+      callback.onSuccess(res, /* httpHeaders */ null);
+    } catch (IOException e) {
+      GoogleJsonResponseException jsonException = ApiErrorExtractor.getJsonResponseException(e);
+      if (jsonException == null) {
+        throw e;
```

Comment:
If the "jsonException" is null, essentially calling onFailure will be fail as we do not have getDetails() or getHeader
(). We will not be able to handle such failure as we do not know if the error is because it was not able to find the folder or any detail. How do you suggest we handle this case or we keep this case altogether as I wasn't able to see from integration testing.